torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py

```diff
@@ -291,6 +298,27 @@ def aot_dispatch_autograd_graph(
 
     fx_g = _create_graph(joint_fn_to_trace, updated_joint_inputs, aot_config=aot_config)
 
+    # See Note [Side-Effectful Tokens in AOTAutograd]
+    # If forward graph does not use effectful ops with Tokens, we have not added additional token arguments before final tracing.
+    # If backward graph does use effectful ops with Tokens,
+    # we have to add additional token argument to the joint graph manually after tracing.
+    if fw_metadata.num_backward_discovered_tokens > 0:
+        add_discovered_token_in_backward_as_input(
```

Comment:
Hmm. Extra graph pass here definitely works, although in general I am a bit hesitant about adding extra graph passes in AOTDispatcher (my mental model today is that our steps are (1) metadata collection, (2) trace a joint graph, (3) partition, (4) compile. And that the graph we got right after joint tracing might not be "correct" and is fixed up elsewhere feels nonlocal/hard to reason about).

I was imagining that instead of a fresh graph pass, we would add the new tokens as extra arguments **at the same time that we are in the middle of tracing**. 

One way of doing this is (which on second thought, might take some... experimentation) would be to update the `handle_effects` ([code](https://github.com/pytorch/pytorch/blob/32e7d284bdbafa6a2ca50b8b9519b5b3e09bdeae/torch/_higher_order_ops/effects.py#L197)) (called by`FunctionalTensor.__torch_dispatch__()`. If we are currently proxy-tracing while this code is running, then do some graph surgery on the fly to add the newly created token as an extra primal. Something like:
```
maybe_mode = torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.PROXY)
if maybe_mode:
    # we are in the middle of the tracing phase of AOTDispatcher
    graph  = maybe_mode.tracer.graph
    ...
    # grab the `primals` placeholder node off of `graph` and add the token tensor as an extra graph arg,
    # similar to what you did in `add_discovered_token_in_backward_as_input`
    # but only need to do for a single token
    # probably also need to mess with the pytree stuff...
    from torch.fx.experimental.proxy_tensor import track_tensor_tree, disable_proxy_modes_tracing
    with disable_proxy_modes_tracing():
         token = torch.tensor([])
     # hook up the newly created token to the currently-traced-graph's list of proxies
     token_proxy = maybe_m.tracer.create_proxy("placeholder", "primals_N", (), {}, name="primals_N")
     track_tensor_tree(token, token_proxy, constant=None, tracer=maybe_m.tracer)
```

Happy to chat more about this next week though

Reply:
Removed the pass, but we still need num_backward_discovered_tokens to add additional arguments at runtime for aot_eager.